rllib/examples/custom_eval.py

```diff
@@ -0,0 +1,196 @@
+------------------------------------------------------------------------
+
+INFO trainer.py:631 -- Running custom eval function <function ...>
+Update corridor length to 4
+Update corridor length to 7
+Custom evaluation round 1
+Custom evaluation round 2
+Custom evaluation round 3
+Custom evaluation round 4
+
+Result for PG_SimpleCorridor_0de4e686:
+  ...
+  evaluation:
+    custom_metrics: {}
+    episode_len_mean: 9.15695067264574
+    episode_reward_max: 1.0
+    episode_reward_mean: 0.9596412556053812
+    episode_reward_min: 0.0
+    episodes_this_iter: 223
+    foo: 1
+    off_policy_estimator: {}
+    policy_reward_max: {}
+    policy_reward_mean: {}
+    policy_reward_min: {}
+    sampler_perf:
+      mean_env_wait_ms: 0.03423667269562796
+      mean_inference_ms: 0.5654563161491506
+      mean_processing_ms: 0.14494765630060774
+"""
+
+import argparse
+import numpy as np
+import gym
+from gym.spaces import Discrete, Box
+
+import ray
+from ray import tune
+from ray.rllib.evaluation.metrics import collect_episodes, summarize_episodes
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--custom-eval", action="store_true")
+args = parser.parse_args()
+
+
+def custom_eval_function(trainer, eval_workers):
+    """Example of a custom evaluation function.
+
+    Arguments:
+        trainer (Trainer): trainer class to evaluate.
+        eval_workers (WorkerSet): evaluation workers.
+
+    Returns:
+        metrics (dict): evaluation metrics dict.
+    """
+
+    # We configured 2 eval workers in the training config.
+    worker_1, worker_2 = eval_workers.remote_workers()
+
+    # Set different env settings for each worker. Here we use a fixed config,
+    # which also could have been computed in each worker by looking at
+    # env_config.worker_index (printed in SimpleCorridor class above).
+    worker_1.foreach_env.remote(lambda env: env.set_corridor_length(4))
+    worker_2.foreach_env.remote(lambda env: env.set_corridor_length(7))
+
+    for i in range(5):
+        print("Custom evaluation round", i)
+        # Calling .sample() runs exactly one episode per worker due to how the
+        # eval workers are configured.
+        ray.get([w.sample.remote() for w in eval_workers.remote_workers()])
+
+    # Collect the accumulated episodes on the workers, and then summarize the
+    # episode stats into a metrics dict.
+    episodes, _ = collect_episodes(
+        remote_workers=eval_workers.remote_workers(), timeout_seconds=99999)
+    # You can compute metrics from the episodes manually, or use the
+    # convenient `summarize_episodes()` utility:
+    metrics = summarize_episodes(episodes)
+    # Note that the above two statements are the equivalent of:
+    # metrics = collect_metrics(eval_workers.local_worker(),
+    #                           eval_workers.remote_workers())
+
+    # You can also put custom values in the metrics dict.
+    metrics["foo"] = 1
+    return metrics
+
+
+class SimpleCorridor(gym.Env):
```

Comment:
Should we share this Env between different examples?

Reply:
It's used in custom_env.py (minus the `set_corridor_length`).

Reply:
I think in this case it's worth it to copy it, which makes the example more self contained. Otherwise, you couldn't run this file standalone.

Reply:
Fair enough. I guess in the `examples` folder, it makes sense to have self-contained code.